# Uvod v NumPy


## Understanding Data Types in Python



```C
/* C code */
int result = 0;
for(int i=0; i<100; i++){
    result += i;
}
```


```python
# Python code
result = 0
for i in range(100):
    result += i
```


### A Python Integer Is More Than Just an Integer



```C
struct _longobject {
    long ob_refcnt;
    PyTypeObject *ob_type;
    size_t ob_size;
    long ob_digit[1];
};
```



<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/cint_vs_pyint.png" alt="Integer Memory Layout">

### A Python List Is More Than Just a List



In [1]:
L = list(range(10))
L

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/array_vs_list.png" alt="Array Memory Layout">

### Fixed-Type Arrays in Python



## How Vectorization Makes Code Faster


<p><img alt="Translating Python code to bytecode" src="https://s3.amazonaws.com/dq-content/289/bytecode.svg"></p>


<table>
<thead>
<tr>
<th>Language Type</th>
<th>Example</th>
<th>Time taken to write program</th>
<th>Control over program performance</th>
</tr>
</thead>
<tbody>
<tr>
<td>High-Level</td>
<td>Python</td>
<td>Low</td>
<td>Low</td>
</tr>
<tr>
<td>Low-Level</td>
<td>C</td>
<td>High</td>
<td>High</td>
</tr>
</tbody>
</table>



<p><img alt="For loop to sum rows" src="https://s3.amazonaws.com/dq-content/289/for_loop.svg"></p>

In [2]:
my_numbers = [[6,5], [1,3], [5,6]]

sums = []

for row in my_numbers:
    row_sum = row[0] + row[1]
    sums.append(row_sum)
    
print(sums)    

[11, 4, 11]




<p><img alt="Unvectorized operation" src="https://s3.amazonaws.com/dq-content/289/unvectorized.svg"></p>


<p><img alt="Vectorized operation" src="https://s3.amazonaws.com/dq-content/289/vectorized.svg"></p>



## Numpy library

[Dokumentacija](http://www.numpy.org/)

In [3]:
import numpy as np #velja skozi cel jupyter notebook

## Introduction to Ndarrays

<img alt="Dimensional Arrays" src="./images/one_dim.svg">

In [4]:
data_ndarray = np.array ([5,10,15,20])

In [5]:
data_ndarray

array([ 5, 10, 15, 20])

In [6]:
type(data_ndarray)

numpy.ndarray

<img alt="Dimensional Arrays" src="./images/Two_Dim.svg">

In [7]:
data_ndarray = np.array ([5,10],[15,20])

TypeError: data type not understood

## Priprava podatkov za delo

<div>

<ul>
<li><code>pickup_year</code>: The year of the trip.</li>
<li><code>pickup_month</code>: The month of the trip (January is <code>1</code>, December is <code>12</code>).</li>
<li><code>pickup_day</code>: The day of the month of the trip.</li>
<li><code>pickup_location_code</code>: The airport or <a target="_blank" href="https://en.wikipedia.org/wiki/Boroughs_of_New_York_City">borough</a> where the trip started.</li>
<li><code>dropoff_location_code</code>: The airport or borough where the trip finished.</li>
<li><code>trip_distance</code>: The distance of the trip in miles.</li>
<li><code>trip_length</code>: The length of the trip in seconds.</li>
<li><code>fare_amount</code>: The base fare of the trip, in dollars.</li>
<li><code>total_amount</code>: The total amount charged to the passenger, including all fees, tolls and tips.</li>
</ul>

</div>

    pickup_year,pickup_month,pickup_day,pickup_dayofweek
    2016,1,1,5
    2016,1,1,5
    2016,1,1,5
    2016,1,1,5

In [ ]:
!head -n 7 ./data/nyc_taxis.csv

    # our list of lists is stored as data_list
    data_ndarray = np.array(data_list)

In [ ]:
import csv
import numpy as np

with open ("./data/nyc_taxis.csv", "r") as f:
    taxi_list = list(csv.reader(f))



In [ ]:
taxi_list = taxi_list [1:] #prvo vrstico oz. header vržemo stran

In [ ]:
converted_taxi_list = [] #spravimo vse skup v float
for row in taxi_list:
    converted_row = []
    for item in row:
        converted_row.append(float(item))
    converted_taxi_list.append(converted_row)
    

In [ ]:
conv = np.array(converted_taxi_list)
conv.shape

In [ ]:
conv.size

In [ ]:
conv.ndim

## Array Shapes

In [10]:
data2 = [[1,2,3,4],[1,2,3,4]]

In [11]:
data2 = np.array(data2)

In [12]:
data2.shape #kakšna je oblika

(2, 4)

In [13]:
conv.itemsize #koliko elementov

NameError: name 'conv' is not defined

In [14]:
conv.nbytes/1024/1024

NameError: name 'conv' is not defined

<div class="alert alert-block alert-info">
<b>Vaja:</b> Assign the shape of taxi to taxi_shape. Print the result.</div>

## Selecting and Slicing Rows and Items from ndarrays

<img alt="Dimensional Arrays" src="./images/selection_rows.svg">

    ndarray[row_index,column_index]

    # or if you want to select all
    # columns for a given set of rows
    ndarray[row_index]

<img alt="Dimensional Arrays" src="./images/selection_item.svg">

In [15]:
test = np.random.randint(0,10, (5,5))

In [16]:
test

array([[9, 3, 5, 2, 4],
       [7, 6, 8, 8, 1],
       [6, 7, 7, 8, 1],
       [5, 9, 8, 9, 4],
       [3, 0, 3, 5, 0]])

In [17]:
test[0]

array([9, 3, 5, 2, 4])

In [18]:
test[-1]

array([3, 0, 3, 5, 0])

In [19]:
test[1:3]

array([[7, 6, 8, 8, 1],
       [6, 7, 7, 8, 1]])

In [21]:
row2_3 = test[1:3]
row2_3

array([[7, 6, 8, 8, 1],
       [6, 7, 7, 8, 1]])

In [22]:
test[2:]

array([[6, 7, 7, 8, 1],
       [5, 9, 8, 9, 4],
       [3, 0, 3, 5, 0]])

In [23]:
test[2,4]

1

In [24]:
taxi

NameError: name 'taxi' is not defined

In [31]:
import csv
import numpy as np

with open ("./data/nyc_taxis.csv", "r") as f:
    taxi_list = list(csv.reader(f))

taxi_list = taxi_list [1:]
converted_taxi_list = [] #spravimo vse skup v float
for row in taxi_list:
    converted_row = []
    for item in row:
        converted_row.append(float(item))
        
    converted_taxi_list.append(converted_row)
conv = np.array(converted_taxi_list)

<div class="alert alert-block alert-info">
<b>Vaja:</b> From the taxi ndarray:
- Select the row at index 0. Assign it to row_0.
- Select every column for the rows at indexes 391 to 500 inclusive. Assign them to rows_391_to_500.
- Select the item at row index 21 and column index 5. Assign it to row_21_column_5.</div>

In [34]:
converted_taxi_list

[[2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  4.0,
  21.0,
  2037.0,
  52.0,
  0.8,
  5.54,
  11.65,
  69.99,
  1.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  1.0,
  16.29,
  1520.0,
  45.0,
  1.3,
  0.0,
  8.0,
  54.3,
  1.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  6.0,
  12.7,
  1462.0,
  36.5,
  1.3,
  0.0,
  0.0,
  37.8,
  2.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  6.0,
  8.7,
  1210.0,
  26.0,
  1.3,
  0.0,
  5.46,
  32.76,
  1.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  6.0,
  5.56,
  759.0,
  17.5,
  1.3,
  0.0,
  0.0,
  18.8,
  2.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  4.0,
  2.0,
  21.45,
  2004.0,
  52.0,
  0.8,
  0.0,
  52.8,
  105.6,
  1.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  6.0,
  8.45,
  927.0,
  24.5,
  1.3,
  0.0,
  6.45,
  32.25,
  1.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  6.0,
  7.3,
  731.0,
  21.5,
  1.3,
  0.0,
  0.0,
  22.8,
  2.0],
 [2016.0,
  1.0,
  1.0,
  5.0,
  0.0,
  2.0,
  5.0,
  36.3,
  2562.0,
  1

In [27]:
taxi_list

[['pickup_year',
  'pickup_month',
  'pickup_day',
  'pickup_dayofweek',
  'pickup_time',
  'pickup_location_code',
  'dropoff_location_code',
  'trip_distance',
  'trip_length',
  'fare_amount',
  'fees_amount',
  'tolls_amount',
  'tip_amount',
  'total_amount',
  'payment_type'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '2',
  '4',
  '21.00',
  '2037',
  '52.00',
  '0.80',
  '5.54',
  '11.65',
  '69.99',
  '1'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '2',
  '1',
  '16.29',
  '1520',
  '45.00',
  '1.30',
  '0.00',
  '8.00',
  '54.30',
  '1'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '2',
  '6',
  '12.70',
  '1462',
  '36.50',
  '1.30',
  '0.00',
  '0.00',
  '37.80',
  '2'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '2',
  '6',
  '8.70',
  '1210',
  '26.00',
  '1.30',
  '0.00',
  '5.46',
  '32.76',
  '1'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '2',
  '6',
  '5.56',
  '759',
  '17.50',
  '1.30',
  '0.00',
  '0.00',
  '18.80',
  '2'],
 ['2016',
  '1',
  '1',
  '5',
  '0',
  '4',
  '2',
 

In [35]:
taxi

NameError: name 'taxi' is not defined

In [36]:
import numpy as np

import csv
import numpy as np

with open('data/nyc_taxis.csv', 'r') as f:
    taxi_list = list(csv.reader(f))
    
taxi_list = taxi_list[1:]

converted_taxi_list = []

for row in taxi_list:
    converted_row = []
    for item in row:
        converted_row.append(float(item))
    converted_taxi_list.append(converted_row)
    
taxi = np.array(converted_taxi_list)

In [38]:
rows_391_to_500 = taxi[391:501]
rows_391_to_500

array([[2.016e+03, 1.000e+00, 2.000e+00, ..., 0.000e+00, 2.630e+01,
        2.000e+00],
       [2.016e+03, 1.000e+00, 2.000e+00, ..., 3.000e+00, 3.030e+01,
        1.000e+00],
       [2.016e+03, 1.000e+00, 2.000e+00, ..., 6.670e+00, 4.001e+01,
        1.000e+00],
       ...,
       [2.016e+03, 1.000e+00, 2.000e+00, ..., 4.960e+00, 2.976e+01,
        1.000e+00],
       [2.016e+03, 1.000e+00, 2.000e+00, ..., 0.000e+00, 3.284e+01,
        2.000e+00],
       [2.016e+03, 1.000e+00, 2.000e+00, ..., 7.050e+00, 4.239e+01,
        1.000e+00]])

## Selecting Columns and Custom Slicing ndarrays

<img alt="Dimensional Arrays" src="./images/selection_columns_updated.svg">

<img alt="Dimensional Arrays" src="./images/selection_1darray_updated.svg">

<img alt="Dimensional Arrays" src="./images/selection_2darray_updated.svg">

In [39]:
column_test = np.random.randint(0, 10, (5,5))

In [40]:
column_test

array([[2, 3, 8, 1, 3],
       [3, 3, 7, 0, 1],
       [9, 9, 0, 4, 7],
       [3, 2, 7, 2, 0],
       [0, 4, 5, 5, 6]])

In [41]:
column_test[:, 3]

array([1, 0, 4, 2, 5])

In [43]:
column_test[:, 1:3]

array([[3, 8],
       [3, 7],
       [9, 0],
       [2, 7],
       [4, 5]])

In [45]:
column_test[:, [1,3,4]]

array([[3, 1, 3],
       [3, 0, 1],
       [9, 4, 7],
       [2, 2, 0],
       [4, 5, 6]])

In [46]:
column_test[2,1:4]

array([9, 0, 4])

In [47]:
column_test[1:, 4]

array([1, 7, 0, 6])

In [49]:
column_test[1:4, :3]

array([[3, 3, 7],
       [9, 9, 0],
       [3, 2, 7]])

<div class="alert alert-block alert-info">
<b>Vaja:</b> From the taxi ndarray:
- Select every row for the columns at indexes 1, 4, and 7. Assign them to columns_1_4_7.
- Select the columns at indexes 5 to 8 inclusive for the row at index 99. Assign them to row_99_columns_5_to_8.
- Select the rows at indexes 100 to 200 inclusive for the column at index 14. Assign them to rows_100_to_200_column_14.</div>

In [52]:
columns_1_4_7 = taxi[:, [1,4,7]]

In [54]:
row_99_columns_5_to_8 =taxi[99, 5:9]

In [55]:
rows_100_to_200_column_14 =taxi[100:201, 14]

## Vector Math

In [56]:
my_numbers =[[6,5], [9,1], [2,4], [7,14], [8,6]]

In [57]:
sums = []
for row in my_numbers:
    row_sums = row[0] + row[1]
    sums.append(row_sums)
    
print(sums)

[11, 10, 6, 21, 14]


In [58]:
my_numbers = np.array(my_numbers) #pretvoraba v numpy array

In [60]:
sums = my_numbers[:,0] + my_numbers[:,1] #seštevanje istoležnih elementov
sums

array([11, 10,  6, 21, 14])

<div class="alert alert-block alert-info">
<b>Vaja:</b> 
Use vector addition to add fare_amount and fees_amount. Assign the result to fare_and_fees.
After you have run your code, use the variable inspector below the code box to inspect the variables.</div>

In [65]:
#taxi stolpec index 9
fare_amount = taxi[:,9]
#stolpec z indeksom 10 
fees_amount = taxi[:,10]
fare_and_fees = fare_amount + fees_amount
fare_and_fees

array([52.8, 46.3, 37.8, ..., 52.8, 35.8, 49.3])

In [67]:
trip_distance = taxi[:,7] #distance
trip_length_seconds = taxi [:,8] #čas potovanja

trip_length_hours = trip_length_seconds / 3600
trip_mph = trip_distance / trip_length_hours
trip_mph

array([37.11340206, 38.58157895, 31.27222982, ..., 22.29907867,
       42.41551247, 36.90473407])



<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The following table lists the arithmetic operators implemented in NumPy:</p>
<table>
<thead><tr>
<th>Operator</th>
<th>Equivalent ufunc</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>+</code></td>
<td><code>np.add</code></td>
<td>Addition (e.g., <code>1 + 1 = 2</code>)</td>
</tr>
<tr>
<td><code>-</code></td>
<td><code>np.subtract</code></td>
<td>Subtraction (e.g., <code>3 - 2 = 1</code>)</td>
</tr>
<tr>
<td><code>-</code></td>
<td><code>np.negative</code></td>
<td>Unary negation (e.g., <code>-2</code>)</td>
</tr>
<tr>
<td><code>*</code></td>
<td><code>np.multiply</code></td>
<td>Multiplication (e.g., <code>2 * 3 = 6</code>)</td>
</tr>
<tr>
<td><code>/</code></td>
<td><code>np.divide</code></td>
<td>Division (e.g., <code>3 / 2 = 1.5</code>)</td>
</tr>
<tr>
<td><code>//</code></td>
<td><code>np.floor_divide</code></td>
<td>Floor division (e.g., <code>3 // 2 = 1</code>)</td>
</tr>
<tr>
<td><code>**</code></td>
<td><code>np.power</code></td>
<td>Exponentiation (e.g., <code>2 ** 3 = 8</code>)</td>
</tr>
<tr>
<td><code>%</code></td>
<td><code>np.mod</code></td>
<td>Modulus/remainder (e.g., <code>9 % 4 = 1</code>)</td>
</tr>
</tbody>
</table>

</div>
</div>


In [70]:
trip_mph1 = np.divide(trip_distance,trip_length_hours)
trip_mph1

array([37.11340206, 38.58157895, 31.27222982, ..., 22.29907867,
       42.41551247, 36.90473407])

## Calculating Statistics For 1D ndarrays

In [71]:
trip_mph.min() #minimalna vrednost

0.0

In [72]:
trip_mph.max() #največja vrednost 

82800.0

In [73]:
trip_mph.mean() #srednja vrednost lahko zapišemo tudi np.mean(trip_mph)

32.24258580925573

In [74]:
trip_mph.sum() #vsota

2887645.985076943


<p></p><center><img alt="Method syntax" src="https://s3.amazonaws.com/dq-content/289/Method_syntax.svg"></center><p></p>


<div class="alert alert-block alert-info">
<b>Vaja:</b> Use the ndarray.max() method to calculate the maximum value of trip_mph. Assign the result to mph_max.
Use the ndarray.mean() method to calculate the average value of trip_mph. Assign the result to mph_mean.</div>

In [76]:
mph_max = trip_mph.max()
mph_mean = trip_mph.mean()
print(mph_max)
print(mph_mean)

82800.0
32.24258580925573


<div>

<table>
<thead>
<tr>
<th>Calculation</th>
<th>Function Representation</th>
<th>Method Representation</th>
</tr>
</thead>
<tbody>
<tr>
<td>Calculate the minimum value of <code>trip_mph</code></td>
<td><code>np.min(trip_mph)</code></td>
<td><code>trip_mph.min()</code></td>
</tr>
<tr>
<td>Calculate the maximum value of <code>trip_mph</code></td>
<td><code>np.max(trip_mph)</code></td>
<td><code>trip_mph.max()</code></td>
</tr>
<tr>
<td>Calculate the <a target="_blank" href="https://en.wikipedia.org/wiki/Mean">mean average</a> value of <code>trip_mph</code></td>
<td><code>np.mean(trip_mph)</code></td>
<td><code>trip_mph.mean()</code></td>
</tr>
<tr>
<td>Calculate the <a target="_blank" href="https://en.wikipedia.org/wiki/Median">median average</a> value of <code>trip_mph</code></td>
<td><code>np.median(trip_mph)</code></td>
<td>There is no ndarray median method</td>
</tr>
</tbody>
</table>
</div>

## Calculating Statistics For 2D ndarrays

<img alt="Dimensional Arrays" src="./images/array_method_axis_none.svg">

<img alt="Dimensional Arrays" src="./images/array_method_axis_1.svg">

<img alt="Dimensional Arrays" src="./images/array_method_axis_0.svg">



<p><img alt="The axis parameter" src="https://s3.amazonaws.com/dq-content/289/axis_param.svg"></p>


In [78]:
test =np.random.randint(0,9, (5,5,))
test

array([[3, 7, 5, 5, 0],
       [1, 5, 3, 0, 5],
       [0, 1, 2, 4, 2],
       [0, 3, 2, 0, 7],
       [5, 0, 2, 7, 2]])

In [79]:
test.max()

7

In [80]:
test.max(axis=0)

array([5, 7, 5, 7, 7])

In [81]:
test.max(axis=1)

array([7, 5, 4, 7, 7])

In [82]:
taxi_first_five = taxi[:5]
fare_components = taxi_first_five[:, 9:13]
fare_components

array([[52.  ,  0.8 ,  5.54, 11.65],
       [45.  ,  1.3 ,  0.  ,  8.  ],
       [36.5 ,  1.3 ,  0.  ,  0.  ],
       [26.  ,  1.3 ,  0.  ,  5.46],
       [17.5 ,  1.3 ,  0.  ,  0.  ]])

In [83]:
fare_components.sum(axis = 1)

array([69.99, 54.3 , 37.8 , 32.76, 18.8 ])

In [84]:
taxi_first_five[:,13]

array([69.99, 54.3 , 37.8 , 32.76, 18.8 ])

## Reading CSV files with NumPy

<p>Below is information about selected columns from the data set:</p>
<ul>
<li><code>pickup_year</code>: The year of the trip.</li>
<li><code>pickup_month</code>: The month of the trip (January is <code>1</code>, December is <code>12</code>).</li>
<li><code>pickup_day</code>: The day of the month of the trip.</li>
<li><code>pickup_location_code</code>: The airport or <a target="_blank" href="https://en.wikipedia.org/wiki/Boroughs_of_New_York_City">borough</a> where the the trip started.</li>
<li><code>dropoff_location_code</code>: The airport or borough where the the trip finished.</li>
<li><code>trip_distance</code>: The distance of the trip in miles.</li>
<li><code>trip_length</code>: The length of the trip in seconds.</li>
<li><code>fare_amount</code>: The base fare of the trip, in dollars.</li>
<li><code>total_amount</code>: The total amount charged to the passenger, including all fees, tolls and tips.</li>
</ul>


In [86]:
taxi = np.genfromtxt('data/nyc_taxis.csv', delimiter =",")

In [87]:
taxi_shape = taxi.shape

In [88]:
taxi_shape

(89561, 15)

In [89]:
taxi.dtype

dtype('float64')

In [90]:
print(taxi) #nan = not a number, ker numpy lahko hrani samo en tip podatkov

[[      nan       nan       nan ...       nan       nan       nan]
 [2.016e+03 1.000e+00 1.000e+00 ... 1.165e+01 6.999e+01 1.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 ... 8.000e+00 5.430e+01 1.000e+00]
 ...
 [2.016e+03 6.000e+00 3.000e+01 ... 5.000e+00 6.334e+01 1.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 8.950e+00 4.475e+01 1.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 0.000e+00 5.484e+01 2.000e+00]]


In [92]:
taxi = np.genfromtxt('data/nyc_taxis.csv', delimiter =",", skip_header=1) #spustimo header vrstico
print (taxi)

[[2.016e+03 1.000e+00 1.000e+00 ... 1.165e+01 6.999e+01 1.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 ... 8.000e+00 5.430e+01 1.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 ... 0.000e+00 3.780e+01 2.000e+00]
 ...
 [2.016e+03 6.000e+00 3.000e+01 ... 5.000e+00 6.334e+01 1.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 8.950e+00 4.475e+01 1.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 0.000e+00 5.484e+01 2.000e+00]]


In [93]:
taxi.shape

(89560, 15)

## Datatypes

In [94]:
x = np.array([1,2])
print(x.dtype)
print(x.nbytes)

int64
16


In [95]:
x = np.array([1.0,2.0])
print(x.dtype)
print(x.nbytes)

float64
16


In [96]:
x = np.array([1,2], dtype = np.int32) #definiranje tipa podatka
print(x.dtype)
print(x.nbytes)

int32
8


In [98]:
x = np.array([1,2], dtype = "int8") #definiranje tipa podatka drug način
print(x.dtype)
print(x.nbytes)

int8
2


In [100]:
x = np.array([189, 22, 128, -129], dtype = 'int8') #rollover
print(x)

[ -67   22 -128  127]


<div class="text_cell_render border-box-sizing rendered_html">
<table>
<thead><tr>
<th>Data type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>bool_</code></td>
<td>Boolean (True or False) stored as a byte</td>
</tr>
<tr>
<td><code>int_</code></td>
<td>Default integer type (same as C <code>long</code>; normally either <code>int64</code> or <code>int32</code>)</td>
</tr>
<tr>
<td><code>intc</code></td>
<td>Identical to C <code>int</code> (normally <code>int32</code> or <code>int64</code>)</td>
</tr>
<tr>
<td><code>intp</code></td>
<td>Integer used for indexing (same as C <code>ssize_t</code>; normally either <code>int32</code> or <code>int64</code>)</td>
</tr>
<tr>
<td><code>int8</code></td>
<td>Byte (-128 to 127)</td>
</tr>
<tr>
<td><code>int16</code></td>
<td>Integer (-32768 to 32767)</td>
</tr>
<tr>
<td><code>int32</code></td>
<td>Integer (-2147483648 to 2147483647)</td>
</tr>
<tr>
<td><code>int64</code></td>
<td>Integer (-9223372036854775808 to 9223372036854775807)</td>
</tr>
<tr>
<td><code>uint8</code></td>
<td>Unsigned integer (0 to 255)</td>
</tr>
<tr>
<td><code>uint16</code></td>
<td>Unsigned integer (0 to 65535)</td>
</tr>
<tr>
<td><code>uint32</code></td>
<td>Unsigned integer (0 to 4294967295)</td>
</tr>
<tr>
<td><code>uint64</code></td>
<td>Unsigned integer (0 to 18446744073709551615)</td>
</tr>
<tr>
<td><code>float_</code></td>
<td>Shorthand for <code>float64</code>.</td>
</tr>
<tr>
<td><code>float16</code></td>
<td>Half precision float: sign bit, 5 bits exponent, 10 bits mantissa</td>
</tr>
<tr>
<td><code>float32</code></td>
<td>Single precision float: sign bit, 8 bits exponent, 23 bits mantissa</td>
</tr>
<tr>
<td><code>float64</code></td>
<td>Double precision float: sign bit, 11 bits exponent, 52 bits mantissa</td>
</tr>
<tr>
<td><code>complex_</code></td>
<td>Shorthand for <code>complex128</code>.</td>
</tr>
<tr>
<td><code>complex64</code></td>
<td>Complex number, represented by two 32-bit floats</td>
</tr>
<tr>
<td><code>complex128</code></td>
<td>Complex number, represented by two 64-bit floats</td>
</tr>
</tbody>
</table>

</div>

## Boolean Indexing

### Boolean Arrays

In [101]:
c= np.array ([80.0, 103.4, 96.9, 200.3])

In [102]:
c_bool= c>100

In [103]:
c_bool

array([False,  True, False,  True])

In [105]:
np.array([2,4,5,9]) + 10

array([12, 14, 15, 19])

In [104]:
result = c[c_bool]
result

array([103.4, 200.3])

In [112]:
month = taxi[:,1]
january_bool = month == 1
january = month[january_bool]
np.size(january)
january.shape[0] #enakovredno

13481

In [114]:
pickup_month = taxi[:,1]
rides = []

for month in range(1,13):
    month_bool = pickup_month == month
    month_array = pickup_month[month_bool]
    month_rides = month_array.shape[0]
    rides.append(month_rides)

print(rides)

[13481, 13333, 15547, 14810, 16650, 15739, 0, 0, 0, 0, 0, 0]


<div class="alert alert-block alert-info">
Use vectorized boolean operations to:
<li> Evaluate whether the elements in array a are less than 3. Assign the result to a_bool.</li> 
<li> Evaluate whether the elements in array b are equal to "blue". Assign the result to b_bool.</li> 
<li>  Evaluate whether the elements in array c are greater than 100. Assign the result to c_bool.</li> </div>

### Boolean Indexing with 1D ndarrays

<div class="alert alert-block alert-info">
Calculate the number of rides in the taxi ndarray that are from February:
<li> Create a boolean array, february_bool, that evaluates whether the items in pickup_month are equal to 2.</li> 
<li> Use the february_bool boolean array to index pickup_month. Assign the result to february.</li> 
<li> Use the ndarray.shape attribute to find the number of items in february. Assign the result to february_rides.</li> </div>

### Boolean Indexing with 2D ndarrays

<img alt="Dimensional Arrays" src="./images/bool_dims_updated.svg">

In [115]:
trip_mph = taxi[:,7] / (taxi[:,8] /3600)
trip_mph.max()

82800.0

In [116]:
trips_over_20000 = taxi[trip_mph > 20000, 5:9]
trips_over_20000

array([[ 2. ,  2. , 23. ,  1. ],
       [ 2. ,  2. , 19.6,  1. ],
       [ 2. ,  2. , 16.7,  2. ],
       [ 3. ,  3. , 17.8,  2. ],
       [ 2. ,  2. , 17.2,  2. ],
       [ 3. ,  3. , 16.9,  3. ],
       [ 2. ,  2. , 27.1,  4. ]])

<div class="alert alert-block alert-info">
<b>Vaja: </b>Ceate a boolean array, tip_bool, that determines which rows have values for the tip_amount column of more than 50. Use the tip_bool array to select all rows from taxi with values tip amounts of more than 50, and the columns from indexes 5 to 13 inclusive. Assign the resulting array to top_tips. </div>

In [119]:
tip_amount = taxi[:,12] 
top_tips = taxi[tip_amount > 50, 5:14]
top_tips[:2] #izpis samo prvih 2 vrstic

array([[4.000e+00, 2.000e+00, 2.145e+01, 2.004e+03, 5.200e+01, 8.000e-01,
        0.000e+00, 5.280e+01, 1.056e+02],
       [3.000e+00, 4.000e+00, 9.200e+00, 1.041e+03, 2.700e+01, 1.300e+00,
        5.540e+00, 6.000e+01, 9.384e+01]])

## Assigning Values

### Assigning Values in ndarrays

In [ ]:
#ndarray[location_ov falues] = values pove kam naj napišemo novo vrednost

In [120]:
a[0] = "orange"
a

array(['orange', 'blue', 'black', 'blue', 'purple'], dtype='<U6')

In [122]:
a = np.array(['red','blue','black','blue','purple'])


In [123]:
a[3:] = "pink"
a

array(['red', 'blue', 'black', 'pink', 'pink'], dtype='<U6')

In [124]:
a[3:] = ["pink", "orange"]
a

array(['red', 'blue', 'black', 'pink', 'orange'], dtype='<U6')

In [126]:
ones = np.ones((3,5))
ones

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [127]:
ones[1,2]=99
ones

array([[ 1.,  1.,  1.,  1.,  1.],
       [ 1.,  1., 99.,  1.,  1.],
       [ 1.,  1.,  1.,  1.,  1.]])

In [129]:
ones[:,2] = 0
ones

array([[1., 1., 0., 1., 1.],
       [1., 1., 0., 1., 1.],
       [1., 1., 0., 1., 1.]])

In [130]:
ones[0] = 42
ones

array([[42., 42., 42., 42., 42.],
       [ 1.,  1.,  0.,  1.,  1.],
       [ 1.,  1.,  0.,  1.,  1.]])

<div class="alert alert-block alert-info">
<b>Vaja: </b>To help you practice without making changes to our original array, we have used the ndarray.copy() method to make taxi_modified, a copy of our original for these exercises.
<li> The value at column index 5 (pickup_location) of row index 28214 is incorrect. Use assignment to change this value to 1 in the taxi_modified ndarray.</li> 
<li> The first column (index 0) contains year values as four digit numbers in the format YYYY (2016, since all trips in our data set are from 2016). Use assignment to change these values to the YY format (16) in the taxi_modified ndarray.</li> 
<li> The values at column index 7 (trip_distance) of rows index 1800 and 1801 are incorrect. Use assignment to change these values in the taxi_modified ndarray to the mean value for that column.</li> </div>

In [131]:
taxi_modified = taxi.copy()

In [132]:
taxi_modified[28215,5] = 1

In [133]:
taxi_modified[:,0] = 16

In [135]:
taxi_modified[1800:1802,7] = taxi_modified[:,7].mean() #nadomesti s povprečno vrednostjo

### Assignment Using Boolean Arrays

In [136]:
a2 = np.array([1,2,3,4,5]) #vsem večjim od 2 priredimo 99
a2[a2>2]=99
a2

array([ 1,  2, 99, 99, 99])

In [137]:
a2 = np.array([1,2,3,4,5]) #vsem večjim od 2 priredimo 99
a2[((a2>2) & (a2<5))] =99
a2

array([ 1,  2, 99, 99,  5])

In [ ]:
# & = and, | = OR, ~ = NOT

In [139]:
taxi_copy = taxi.copy()
total_amount = taxi_copy[:,13]
total_amount[total_amount<0] = 0 # vse vrednosti manjše od 0 postavimo na 0

<div class="alert alert-block alert-info">
<b>Vaja: </b>We again used the ndarray.copy() method to make taxi_copy, a copy of our original for this exercise.
<li> Select the fourteenth column (index 13) in taxi_copy. Assign it to a variable named total_amount.</li> 
<li> For rows where the value of total_amount is less than 0, use assignment to change the value to 0.</li> 
 </div>

In [140]:
b = np.linspace(1,9, num = 9, dtype = np.int)
b

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [141]:
b= np.reshape(b, (3,3))
b

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [142]:
b[b>4]= 99
b

array([[ 1,  2,  3],
       [ 4, 99, 99],
       [99, 99, 99]])

In [145]:
c = np.linspace(1,9, num = 9, dtype = np.int)
c = np.reshape(b, (3,3))
c[c[:,1] >2,1] =99
c

array([[ 1,  2,  3],
       [ 4, 99, 99],
       [99, 99, 99]])

<div class="alert alert-block alert-info">
<b>Vaja: </b>We have created a new copy of our taxi dataset, taxi_modified with an additional column containing the value 0 for every row. In our new column at index 15, assign the value 1 if the pickup_location_code (column index 5) corresponds to an airport location, leaving the value as 0 otherwise by performing these three operations:
<li> For rows where the value for the column index 5 is equal to 2 (JFK Airport), assign the value 1 to column index 15.</li> 
<li>For rows where the value for the column index 5 is equal to 3 (LaGuardia Airport), assign the value 1 to column index 15.</li> 
<li> For rows where the value for the column index 5 is equal to 5 (Newark Airport), assign the value 1 to column index 15.</li> </div>

## Adding Rows and Columns to ndarrays

In [146]:
ones = np.ones((2,3))
ones

array([[1., 1., 1.],
       [1., 1., 1.]])

In [148]:
zeros = np.zeros(3)
zeros

array([0., 0., 0.])

In [149]:
combined = np.concatenate([ones, zeros], axis = 0) #napačne dimenzije

ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)

In [151]:
zeros_2d = np.expand_dims(zeros, axis =0) (iz 3, naredimo 1,3)
zeros_2d

array([[0., 0., 0.]])

In [152]:
zeros_2d.shape

(1, 3)

In [154]:
combined = np.concatenate([ones, zeros_2d], axis = 0) #prave dimnzije
combined

array([[1., 1., 1.],
       [1., 1., 1.],
       [0., 0., 0.]])

## Computation on NumPy Arrays: Universal Functions


### The Slowness of Loops



In [155]:
np.random.seed(0)

def compute_reciprocals(values):
    output = np.empty(len(values))
    for i in range(len(values)):
        output[i] = 1.0 / values[i]
    return output
        
values = np.random.randint(1, 10, size=5)
compute_reciprocals(values)

array([0.16666667, 1.        , 0.25      , 0.25      , 0.125     ])

In [157]:
big_array = np.random.randint(1,100, size = 1000000)
%timeit compute_reciprocals(big_array)

7.29 s ± 956 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Introducing UFuncs (Universal functions)


In [159]:
print(compute_reciprocals(values))
print(1.0 / values)


[0.16666667 1.         0.25       0.25       0.125     ]
[0.16666667 1.         0.25       0.25       0.125     ]


In [167]:
big_array = np.random.randint(1,100, size = 1000000) #full hitreje
%timeit (1.0/big_array)

4.24 ms ± 1.27 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Subarrays as no-copy views



In [165]:
r = np.ones((4,4))
r


array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [166]:
r2 = r[:2, :2]
r2

array([[1., 1.],
       [1., 1.]])

In [168]:
r2[:]= 0 #ko spreminjam vrednosti v podarrayu, se spremenijo tudi vrednosti v osnovnem arrayu.
r2

array([[0., 0.],
       [0., 0.]])

In [169]:
r

array([[0., 0., 1., 1.],
       [0., 0., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

## Copying Data



In [170]:
r = np.ones((4,4))
r3 = r[:2, :2].copy() #ustvarimo kopijo

In [171]:
r

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

## Primer: Which is the most popular airport?

In [173]:
jfk = taxi[taxi[:, 6] == 2]
jfk.shape[0]

11832

In [174]:
laguardia = taxi[taxi[:, 6] == 3]
laguardia.shape[0]

16602

In [175]:
newark = taxi[taxi[:, 6] == 5]
newark.shape[0]

63

## Primer: Calculating Statistics for Trips on Clean Data

In [177]:
trip_mph = taxi[:,7] / (taxi[:,8] / 3600)
cleaned_taxi = taxi[trip_mph < 100]

In [178]:
cleaned_taxi[:, 7].mean() #povprečna razdalja

12.666396599932893

In [180]:
cleaned_taxi[:,8].mean()/60 #povprečna vožnja v minutah

37.325060955150434

In [183]:
cleaned_taxi[:,13].mean() #srednja vrednost

48.98131853260262